apm-agent-plugins/apm-asynchttpclient-plugin/src/main/java/co/elastic/apm/agent/asynchttpclient/AsyncHttpClientInstrumentation.java

```diff
@@ -0,0 +1,153 @@
+import co.elastic.apm.agent.bci.VisibleForAdvice;
+import co.elastic.apm.agent.http.client.HttpClientHelper;
+import co.elastic.apm.agent.impl.transaction.Span;
+import co.elastic.apm.agent.impl.transaction.TraceContext;
+import co.elastic.apm.agent.impl.transaction.TraceContextHolder;
+import io.netty.handler.codec.http.HttpHeaders;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.asynchttpclient.*;
+
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.Collection;
+
+import static net.bytebuddy.matcher.ElementMatchers.*;
+
+public class AsyncHttpClientInstrumentation extends ElasticApmInstrumentation {
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return hasSuperType(named("org.asynchttpclient.AsyncHttpClient"));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("executeRequest")
+            .and(takesArgument(0, named("org.asynchttpclient.Request")))
+            .and(takesArgument(1, named("org.asynchttpclient.AsyncHandler")));
+
+    }
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Arrays.asList("http-client", "asynchttpclient");
+    }
+
+    @Override
+    public Class<?> getAdviceClass() {
+        return AsyncHttpClientInstrumentationAdvice.class;
+    }
+
+    public static class AsyncHttpClientInstrumentationAdvice {
+        @Nullable
+        @VisibleForAdvice
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        public static void onBeforeExecute(@Advice.Argument(value = 0, readOnly = true) Request request,
+                                           @Advice.Argument(value = 1, readOnly = false) AsyncHandler asyncHandler,
+                                           @Advice.Local("span") Span span) {
+            if(tracer == null || tracer.getActive() == null) {
+                return;
+            }
+
+            final TraceContextHolder<?> parent = tracer.getActive();
+            span = HttpClientHelper.startHttpClientSpan(parent, request.getMethod(), request.getUri().toFullUrl(), request.getUri().getHost());
+
+            if(span != null) {
+                span.activate();
+                request.getHeaders().add(TraceContext.TRACE_PARENT_HEADER, span.getTraceContext().getOutgoingTraceParentHeader().toString());
+            }
+            if(!request.getHeaders().contains(TraceContext.TRACE_PARENT_HEADER) && parent != null) {
+                request.getHeaders().add(TraceContext.TRACE_PARENT_HEADER, parent.getTraceContext().getOutgoingTraceParentHeader().toString());
+            }
+
+            // TODO: Optimize the new call away
+            asyncHandler = new AsyncHandlerWrapperCreator().wrap(asyncHandler, span, request);
```

Comment:
This will cause classloader/class visibility issues. The agent code lives within the system/bootstrap class loader, while the HTTP client might live in a Webapp class loader. For example when deploying a war file to tomcat which has the asynchttpclient in the libs folder. In that situation, the agent does not have access to the asynchttpclient classes, as they are loaded in a child classloader. Use `HelperClassManager.ForAnyClassLoader.of` to work around that problem. Take a look at how it's done in `ApacheHttpAsyncClientInstrumentation`.

Reply:
Probably obsolete with https://github.com/elastic/apm-agent-java/pull/622#pullrequestreview-233394323